token/confidential-transfer/ciphertext-arithmetic/src/lib.rs

```diff
@@ -0,0 +1,334 @@
+) -> Option<PodElGamalCiphertext> {
+    let shift_scalar = u64_to_scalar(1_u64 << SHIFT_BITS);
+    let shifted_right_ciphertext_hi = multiply(&shift_scalar, right_ciphertext_hi)?;
+    let combined_right_ciphertext = add(right_ciphertext_lo, &shifted_right_ciphertext_hi)?;
+    subtract(left_ciphertext, &combined_right_ciphertext)
+}
+
+/// Add a constant amount to a ciphertext
+pub fn add_to(ciphertext: &PodElGamalCiphertext, amount: u64) -> Option<PodElGamalCiphertext> {
+    let amount_scalar = u64_to_scalar(amount);
+    let amount_point = multiply_ristretto(&amount_scalar, &G)?;
+
+    let (commitment, handle) = elgamal_ciphertext_to_ristretto(ciphertext);
+
+    let result_commitment = add_ristretto(&commitment, &amount_point)?;
+
+    Some(ristretto_to_elgamal_ciphertext(&result_commitment, &handle))
+}
+
+/// Subtract a constant amount to a ciphertext
+pub fn subtract_from(
+    ciphertext: &PodElGamalCiphertext,
+    amount: u64,
+) -> Option<PodElGamalCiphertext> {
+    let amount_scalar = u64_to_scalar(amount);
+    let amount_point = multiply_ristretto(&amount_scalar, &G)?;
+
+    let (commitment, handle) = elgamal_ciphertext_to_ristretto(ciphertext);
+
+    let result_commitment = subtract_ristretto(&commitment, &amount_point)?;
+
+    Some(ristretto_to_elgamal_ciphertext(&result_commitment, &handle))
+}
+
+/// Convert a `u64` amount into a curve25519 scalar
+fn u64_to_scalar(amount: u64) -> PodScalar {
+    let mut amount_bytes = [0u8; 32];
+    amount_bytes[..8].copy_from_slice(&amount.to_le_bytes());
+    PodScalar(amount_bytes)
+}
+
+/// Convert a `PodElGamalCiphertext` into a tuple of commitment and decrypt
+/// handle `PodRistrettoPoint`
+fn elgamal_ciphertext_to_ristretto(
+    ciphertext: &PodElGamalCiphertext,
+) -> (PodRistrettoPoint, PodRistrettoPoint) {
+    let ciphertext_bytes = bytes_of(ciphertext); // must be of length 64 by type
+    let commitment_bytes = ciphertext_bytes[..32].try_into().unwrap();
+    let handle_bytes = ciphertext_bytes[32..64].try_into().unwrap();
+    (
+        PodRistrettoPoint(commitment_bytes),
+        PodRistrettoPoint(handle_bytes),
+    )
+}
+
+/// Convert a pair of `PodRistrettoPoint` to a `PodElGamalCiphertext`
+/// interpretting the first as the commitment and the second as the handle
+fn ristretto_to_elgamal_ciphertext(
+    commitment: &PodRistrettoPoint,
+    handle: &PodRistrettoPoint,
+) -> PodElGamalCiphertext {
+    let mut ciphertext_bytes = [0u8; 64];
+    ciphertext_bytes[..32].copy_from_slice(bytes_of(commitment));
+    ciphertext_bytes[32..64].copy_from_slice(bytes_of(handle));
+    // Unfortunately, the `solana-zk-sdk` does not exporse a constructor interface
+    // to construct `PodRistrettoPoint` from bytes. As a work-around, encode the
+    // bytes as base64 string and then convert the string to a
+    // `PodElGamalCiphertext`.
+    let ciphertext_string = STANDARD.encode(ciphertext_bytes);
+    FromStr::from_str(&ciphertext_string).unwrap()
```

Comment:
Is it worth adding the constructor in future work? This seems a bit strange

Reply:
Yes, we should add the constructor on the zk-sdk side. I'l create an issue on the SPL side to use the new constructor once it lands.